### PR TITLE
PgValue variant bypassing QValue for more efficient & correct PG<>PG mirroring

### PR DESCRIPTION
**Pull request by serprex**
Currently enabled by passing `system = 'PG'` in `CREATE MIRROR` command

### CODE REVIEW

File: protos/flow.proto [hunk 1]
```diff
@@ -71,6 +71,8 @@ message FlowConnectionConfigs {
   string synced_at_col_name = 19;
 
   string script = 20;
+
+  bool same_same = 21;
```
Comment (serprex) [inline protos/flow.proto:75]: This should be changed to an enum re what type system is used (current, QValue vs PGValue)

**Comment by serprex** [comment]
Problem: table schema deltas don't care about numeric precision/scale

**Comment by serprex** [comment]
Problem: TypeModifier only used for numeric; should be used for all types such as `varchar`

File: flow/connectors/postgres/normalize_stmt_generator.go [hunk 2]
```diff
@@ -98,42 +110,44 @@ func (n *normalizeStmtGenerator) generateFallbackStatements() []string {
...
 	quotedColumnNames := make([]string, columnCount)
 
 	flattenedCastsSQLArray := make([]string, 0, columnCount)
-	parsedDstTable, _ := utils.ParseSchemaTable(n.dstTableName)
+	parsedDstTable, _ := utils.ParseSchemaTable(dstTableName)
 
 	primaryKeyColumnCasts := make(map[string]string)
-	primaryKeySelectSQLArray := make([]string, 0, len(n.normalizedTableSchema.PrimaryKeyColumns))
-	for i, column := range n.normalizedTableSchema.Columns {
+	primaryKeySelectSlice := make([]string, 0, len(normalizedTableSchema.PrimaryKeyColumns))
```
Comment (heavycrystal) [inline flow/connectors/postgres/normalize_stmt_generator.go:133]: nit: SQL in var names used to construct SQL statements later

File: flow/connectors/postgres/postgres.go [hunk 3]
```diff
@@ -308,8 +308,30 @@ func (c *PostgresConnector) SetLastOffset(ctx context.Context, jobName string, l
...
+	ctx context.Context,
+	catalogPool *pgxpool.Pool,
+	req *model.PullRecordsRequest[model.PgWalItems],
+) error {
+	return PullCore(ctx, c, catalogPool, req, (*PostgresCDCSource).PullPgWal)
+}
+
 // PullRecords pulls records from the source.
-func (c *PostgresConnector) PullRecords(ctx context.Context, catalogPool *pgxpool.Pool, req *model.PullRecordsRequest) error {
+func PullCore[T model.Items](
```
Comment (heavycrystal) [inline flow/connectors/postgres/postgres.go:328]: nit: unexport

File: flow/connectors/postgres/postgres.go [hunk 4]
```diff
@@ -838,6 +916,96 @@ func (c *PostgresConnector) SetupNormalizedTable(
 	return false, nil
 }
 
+// TODO share code between SetupNormalizedTable/SetupNormalizedPgTable
```
Comment (heavycrystal) [inline flow/connectors/postgres/postgres.go:919]: yes

File: flow/connectors/postgres/postgres.go [hunk 5]
```diff
@@ -838,6 +916,96 @@ func (c *PostgresConnector) SetupNormalizedTable(
...
+	_, err = createNormalizedTablesTx.Exec(ctx, normalizedTableCreateSQL)
+	if err != nil {
+		return false, fmt.Errorf("error while creating normalized table: %w", err)
+	}
+
+	return false, nil
+}
+
+// TODO share code between ReplayPgTableSchemaDeltas/ReplayTableSchemaDeltas
+// ReplayTableSchemaDelta changes a destination table to match the schema at source
```
Comment (heavycrystal) [inline flow/connectors/postgres/postgres.go:954]: share code, handling the case where we want to add typmod for schema changes

File: flow/connectors/postgres/qvalue_convert.go [hunk 6]
```diff
@@ -18,6 +19,29 @@ import (
...
+	// workaround for some types not being defined by pgtype
+	switch recvOID {
+	case uint32(oid.T_timetz):
+		return "timetz"
+	case uint32(oid.T_xml):
+		return "xml"
+	case uint32(oid.T_money):
+		return "money"
+	case uint32(oid.T_txid_snapshot):
+		return "TXID_SNAPSHOT"
```
Comment (heavycrystal) [inline flow/connectors/postgres/qvalue_convert.go:35]: nit: consistency

File: flow/connectors/postgres/qvalue_convert.go [hunk 7]
```diff
@@ -340,8 +361,10 @@ func parseFieldFromQValueKind(qvalueKind qvalue.QValueKind, value interface{}) (
 			return qvalue.QValueMacaddr{Val: v}, nil
 		case [16]byte:
```
Comment (heavycrystal) [inline flow/connectors/postgres/qvalue_convert.go:362]: check this branch?

File: flow/model/cdc_record_stream.go [hunk 8]
```diff
@@ -7,11 +7,11 @@ import (
 	"github.com/PeerDB-io/peer-flow/peerdbenv"
 )
 
-type CDCRecordStream struct {
+type CDCStream[T Items] struct {
 	// empty signal to indicate if the records are going to be empty or not.
 	emptySignal chan bool
-	// Records are a list of json objects.
-	records chan Record
+	// Records are a list of records.
```
Comment (heavycrystal) [inline flow/model/cdc_record_stream.go:13]: pg wal type or qvalue type

File: flow/model/pg_wal_items.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
+package model
```
Comment (heavycrystal) [inline flow/model/pg_wal_items.go:1]: clean up if needed

File: protos/flow.proto [hunk 10]
```diff
@@ -71,6 +71,8 @@ message FlowConnectionConfigs {
   string synced_at_col_name = 19;
```
Comment (heavycrystal) [inline protos/flow.proto:71]: `TableSchema` uses string regardless of whether it is QValue/PGtype.

Use an enum or similar to help distinguish

**Review by heavycrystal — DISMISSED** [review]

File: flow/activities/flowable.go [hunk 11]
```diff
@@ -1003,34 +1072,19 @@ func (a *FlowableActivity) RenameTables(ctx context.Context, config *protos.Rena
 	*protos.RenameTablesOutput, error,
 ) {
 	ctx = context.WithValue(ctx, shared.FlowNameKey, config.FlowJobName)
-	dstConn, err := connectors.GetCDCSyncConnector(ctx, config.Peer)
+	conn, err := connectors.GetConnectorAs[connectors.RenameTablesConnector](ctx, config.Peer)
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/activities/flowable.go:1075]: #1572

File: flow/connectors/postgres/cdc.go [hunk 12]
```diff
@@ -154,7 +340,7 @@ func (p *PostgresCDCSource) PullRecords(ctx context.Context, req *model.PullReco
 	nextStandbyMessageDeadline := time.Now().Add(standbyMessageTimeout)
 
 	logger := logger.LoggerFromCtx(ctx)
-	addRecordWithKey := func(key *model.TableWithPkey, rec model.Record) error {
+	addRecordWithKey := func(key model.TableWithPkey, rec model.Record[T]) error {
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/connectors/postgres/cdc.go:343]: #1573

File: flow/connectors/postgres/normalize_stmt_generator.go [hunk 13]
```diff
@@ -19,66 +19,78 @@ type normalizeStmtGenerator struct {
...
 	}
-	return n.generateFallbackStatements()
+	return n.generateFallbackStatements(dstTable, normalizedTableSchema)
 }
 
-func (n *normalizeStmtGenerator) generateFallbackStatements() []string {
-	columnCount := len(n.normalizedTableSchema.Columns)
+func (n *normalizeStmtGenerator) generateFallbackStatements(
+	dstTableName string,
+	normalizedTableSchema *protos.TableSchema,
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/connectors/postgres/normalize_stmt_generator.go:60]: #1575

File: flow/connectors/postgres/qvalue_convert.go (hunk 7 above)
Comment (serprex) [reply flow/connectors/postgres/qvalue_convert.go:362]: #1585

**Comment by iskakaushik** [comment]
can you share some examples where text representation of QValue differs from PGValue?

File: flow/activities/flowable.go [hunk 14]
```diff
@@ -542,6 +597,20 @@ func (a *FlowableActivity) StartNormalize(
...
+	input *protos.StartNormalizeInput,
+) (*model.NormalizeResponse, error) {
+	return normalizeCore(ctx, a, input, connectors.CDCNormalizeConnector.NormalizeRecords)
+}
+
+func (a *FlowableActivity) StartNormalizePgWal(
+	ctx context.Context,
+	input *protos.StartNormalizeInput,
+) (*model.NormalizeResponse, error) {
+	return normalizeCore(ctx, a, input, connectors.CDCNormalizePgWalConnector.NormalizePgWal)
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/activities/flowable.go]: disagree, it's confusing when a function is defined in one file & then has its only invocation in another

File: flow/connectors/postgres/cdc.go [hunk 15]
```diff
@@ -349,49 +535,46 @@ func (p *PostgresCDCSource) PullRecords(ctx context.Context, req *model.PullReco
...
 						// No need to store DeleteRecords in memory or disk.
-						err = addRecordWithKey(nil, rec)
+						err = addRecordWithKey(model.TableWithPkey{}, rec)
 						if err != nil {
 							return err
 						}
 					}
 
-				case *model.RelationRecord:
+				case *model.RelationRecord[T]:
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/connectors/postgres/cdc.go]: it does need the type parameter tho, in order to implement generic interface

**Comment by serprex** [comment]
Thoughts/listing things which can be done as followup:

1. types have typmod_in/typmod_out which can be used to properly propagate to destination, need to transmit string instead of raw integer value
2. `not null` & other constraints/indexes, complicates schema changes since not sure what to do about ingesting after a column drop if column is not null, similarly constraints may be violated by deleted columns, so constraints would have to ignore soft deleted columns
4. process type messages in replication stream, adding to custom type map
5. support types without binary serialization (aclitem, seg) which doesn't work with pgx CopyFrom in initial copy. Also think there's issues with custom enums. Need some fallback method
6. TRUNCATE
7. scripting will get all columns as text, transforms will be responsible for preserving valid text representation
8. collations in table schema

File: flow/connectors/postgres/cdc.go [hunk 16]
```diff
@@ -734,6 +806,7 @@ func (p *PostgresCDCSource) processRelationMessage(
 	// tableNameSchemaMapping uses dst table name as the key, so annoying lookup
 	prevSchema := p.tableNameSchemaMapping[p.tableNameMapping[p.srcTableIDNameMapping[currRel.RelationID]].Name]
 	// creating maps for lookup later
+	// TODO don't be qvalue for raw pg
```
Comment (serprex) [inline flow/connectors/postgres/cdc.go:809]: TODO

File: flow/activities/flowable.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (heavycrystal) [reply flow/activities/flowable.go]: IDEs can jump to definitions easily, debate is between ergonomics of multiple files vs one really large file and I prefer the former

**Review by iskakaushik — APPROVED** [review]

**Pull request merged**